File: circle-center/src/api/manager/iconio.ts
import { post, request } from '../client';
import type { GetIconImageResponse, UploadIconResponse } from './types';

// ==========================
// Icon Upload & Retrieval API
// ==========================

/**
 * Upload an icon file for a specific project and component
 * @param projectId - The project ID
 * @param componentInfo - The component info to match with icon record
 * @param file - The icon file to upload
 */
export async function uploadIcon(
  projectId: number,
  componentInfo: string,
  file: File
): Promise<UploadIconResponse> {
  const formData = new FormData();
  formData.append('component_info', componentInfo);
  formData.append('file', file);

  const response = await post<UploadIconResponse>(
    `/manager/icons/${projectId}/upload`,
    formData,
    {
      asFormData: true,
    }
  );

  return response.data;
}

/**
 * Get icon image by relative path (requires authentication)
 * @param relPath - The relative path from uploads root (e.g., "icons/123/my_icon.png")
 */
export async function getIcon(relPath: string): Promise<GetIconImageResponse> {
  try {
    const response = await request<ArrayBuffer>({
      url: `/manager/icons/${relPath}`,
      method: 'GET',
      responseType: 'arraybuffer',
    });

    return {
      success: true,
      message: 'Icon retrieved successfully',
      data: {
        data: response.data,
        content_type:
          response.headers['content-type'] || 'application/octet-stream',
      },
    };
  } catch (error: any) {
    if (error.response?.data?.code === 'ICON_FILE_NOT_FOUND') {
      throw new Error('ICON_FILE_NOT_FOUND');
    }
    throw error;
  }
}

/**
 * Get icon URL for display (creates a blob URL from the icon data)
 * @param relPath - The relative path from uploads root
 */
export async function getIconUrl(relPath: string): Promise<string> {
  try {
    const response = await getIcon(relPath);
    const blob = new Blob([response.data.data], {
      type: response.data.content_type,
    });
    return URL.createObjectURL(blob);
  } catch (error) {
    console.error('Failed to get icon URL:', error);
    throw error;
  }
}

/**
 * Revoke a blob URL to free memory
 * @param url - The blob URL to revoke
 */
export function revokeIconUrl(url: string): void {
  if (url.startsWith('blob:')) {
    URL.revokeObjectURL(url);
  }
}

File: panel/manager/operation/iconio.go
package manager

import (
	"database/sql"
	"io"
	"net/http"
	"os"
	"strconv"
	"strings"

	"github.com/gin-gonic/gin"
	"github.com/h2non/filetype"

	accountsvc "circle-center/panel/account/svc"
	oputils "circle-center/panel/account/utils"
	svc "circle-center/panel/manager/svc"
)

// IconIOHandler exposes HTTP handlers for icon upload and retrieval
type IconIOHandler struct {
	service *svc.IconIOService
}

// NewIconIOHandler constructs a new IconIOHandler
func NewIconIOHandler(db *sql.DB, authClient *accountsvc.AuthClient) *IconIOHandler {
	service, err := svc.NewIconIOService(db, authClient)
	if err != nil {
		panic("Failed to create IconIOService: " + err.Error())
	}
	return &IconIOHandler{service: service}
}

// UploadIcon handles POST /manager/icons/:projectId/upload
// Form fields: component_info (string), file (multipart file)
func (h *IconIOHandler) UploadIcon(c *gin.Context) {
	// Auth: extract token from context or header
	token, ok := oputils.GetTokenFromContext(c)
	if !ok {
		var err error
		token, err = oputils.ExtractBearerToken(c)
		if err != nil {
			oputils.RespondWithAuthError(c, err)
			return
		}
	}

	projectIDStr := c.Param("projectId")
	projectID, err := strconv.ParseUint(strings.TrimSpace(projectIDStr), 10, 64)
	if err != nil {
		c.JSON(http.StatusBadRequest, gin.H{"error": "INVALID_PROJECT_ID", "message": "project id must be uint"})
		return
	}

	componentInfo := strings.TrimSpace(c.PostForm("component_info"))
	if componentInfo == "" {
		c.JSON(http.StatusBadRequest, gin.H{"error": "INVALID_REQUEST", "message": "component_info is required"})
		return
	}

	file, _, err := c.Request.FormFile("file")
	if err != nil {
		c.JSON(http.StatusBadRequest, gin.H{"error": "FILE_REQUIRED", "message": "file is required"})
		return
	}
	defer file.Close()

	data, err := io.ReadAll(file)
	if err != nil {
		c.JSON(http.StatusBadRequest, gin.H{"error": "FILE_READ_ERROR", "message": err.Error()})
		return
	}

	path, err := h.service.ValidateAndSaveIcon(c.Request.Context(), token, projectID, componentInfo, data)
	if err != nil {
		c.JSON(http.StatusBadRequest, gin.H{"error": "UPLOAD_FAILED", "message": err.Error()})
		return
	}

	// Infer content-type for response convenience
	ct := "application/octet-stream"
	if kind, err := filetype.Match(data); err == nil && kind != filetype.Unknown {
		ct = kind.MIME.Value
	}

	c.JSON(http.StatusOK, gin.H{
		"success": true,
		"message": "Icon uploaded successfully",
		"data": gin.H{
			"path":         path,
			"content_type": ct,
		},
	})
}

// GetIcon handles GET /manager/icon/*relpath
// Requires Bearer token; returns the raw bytes with appropriate content-type
func (h *IconIOHandler) GetIcon(c *gin.Context) {
	// Auth
	token, ok := oputils.GetTokenFromContext(c)
	if !ok {
		var err error
		token, err = oputils.ExtractBearerToken(c)
		if err != nil {
			oputils.RespondWithAuthError(c, err)
			return
		}
	}

	rel := c.Param("relpath")
	if len(rel) > 0 && rel[0] == '/' {
		rel = rel[1:]
	}

	abs, err := h.service.GetIconAbsolutePathSecure(c.Request.Context(), token, rel)
	if err != nil {
		c.JSON(http.StatusNotFound, gin.H{"error": "NOT_FOUND", "message": err.Error()})
		return
	}

	if _, err := os.Stat(abs); os.IsNotExist(err) {
		c.JSON(http.StatusNotFound, gin.H{
			"error": "ICON_FILE_NOT_FOUND", 
			"message": "Icon file not uploaded yet",
			"code": "ICON_FILE_NOT_FOUND",
		})
		return
	}

	bytes, err := os.ReadFile(abs)
	if err != nil {
		c.JSON(http.StatusInternalServerError, gin.H{"error": "READ_FAILED", "message": err.Error()})
		return
	}

	ct := "application/octet-stream"
	if kind, err := filetype.Match(bytes); err == nil && kind != filetype.Unknown {
		ct = kind.MIME.Value
	}

	c.Header("Content-Type", ct)
	c.Header("Cache-Control", "private, max-age=31536000")
	c.Status(http.StatusOK)
	_, _ = c.Writer.Write(bytes)
}
